File: langs/scala/runtime/js/src/main/scala/nobleidl/sjs/core/ErrorChecker.scala
package nobleidl.sjs.core

import scala.scalajs.js
import scala.scalajs.js.annotation.{JSImport, JSName}

trait NobleIdlError[Name <: String] extends js.Error {
  @JSName(ErrorChecker.nobleidlErrorTypeSymbol)
  val errorType: Name
  val information: Any
}

trait ErrorChecker[E] extends js.Object {
  def isInstance(x: Any): Boolean
}

@JSImport("@argon-lang/noble-idl-core/util")
@js.native
object ErrorChecker extends js.Object {
  val nobleidlErrorTypeSymbol: js.Symbol = js.native
  
  def fromTypeName[Name <: String, E <: NobleIdlError[Name]](name: Name): ErrorChecker[E] = js.native
}

File: langs/java/runtime/src/main/java/dev/argon/nobleidl/runtime/graaljsInterop/ErrorChecker.java
package dev.argon.nobleidl.runtime.graaljsInterop;

import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Value;

public class ErrorChecker {
	private ErrorChecker() {}

	public static Value forErrorName(Context context, String name) {
		return context.eval("js",
			"""
				name => ({
					isInstance(o) {
						return o[Symbol.for("nobleidl-error-type")] === name;
					},
				})
			"""
		).execute(name);
	}
}
